scripts/runnable_recipe_ray_vllm_weight_sync.py

```diff
@@ -368,6 +369,447 @@ def run(self):
+from tensordict.utils import _zip_strict, expand_as_right
+
+# class Trajectory(TensorClass):
+#     text: str
+#     tokens: torch.Tensor|None=None
+#     response: str | None=None
+#     # response_tokens --> tokens_response because Transform hardcodes..
+#     tokens_response: torch.Tensor | None=None
+#     # prompt_attention_mask: torch.Tensor | None=None
+#     log_probs: torch.Tensor | None=None
+
+def vllm_generate(
+    tensordict: TensorDictBase | Trajectory,
+    vllm_instance,  # The LLM object
+    generate_kwargs: Dict[str, Any] | None = None,
+    tokenizer = None,
+    text_key: str = "text",
+    token_key: str = "tokens",
+    # response_tokens --> tokens_response because Transform hardcodes..
+    token_response_key: str = "tokens_response",
+    text_response_key: str = "text_response",
+    log_prob_key: str = "log_probs",
+    attention_mask_key: str = "attention_mask",
+    padding_value: int = -1,
+) -> Trajectory:
+    
+    args = ()
+    
+    if generate_kwargs is None:
+        generate_kwargs = {}
+    generate_kwargs.setdefault("detokenize", True)
+    generate_kwargs.setdefault("prompt_logprobs", False)
+    generate_kwargs.setdefault("logprobs", True)
+    # Create SamplingParams from generate_kwargs
+    sampling_params = SamplingParams(**generate_kwargs)
+    kwargs = {"sampling_params": sampling_params, "use_tqdm": False}
+
+    
+    txt = tensordict.get(text_key)
+    if not isinstance(txt, (list, str)):
+        txt = txt.tolist()
+    args = (txt,)
+
+    tokens_out = vllm_instance.generate(*args, **kwargs)
+    tokens_out = _get_output_tokens_and_log_probs(tokens_out, tokenizer, log_prob_key, token_response_key, text_response_key, padding_value)
+
+    in_keys = [
+        log_prob_key,
+        token_response_key,
+        text_response_key,
+        token_key,
+        attention_mask_key,
+    ]
+    out = tokens_out.select(*in_keys, strict=False)
+
+    # ================================================================
+    out[token_key] = tensordict[token_key]
+    # ================================================================
+
+    inputs = tensordict.select(text_key, strict=False)
+    if inputs.ndim < out.ndim:
+        inputs = inputs.unsqueeze(-1).expand(out.shape)
+    out.update(inputs)
+
+    td = tensordict.clone()
+    td.update(out, keys_to_update=list(out.keys()))
+
+    # This runs into errors with accessing resut["bla"] e.g. llm.py:511
+    # return Trajectory.from_tensordict(out)
+    return td 
+
+
+CompletionOutput_tc = from_dataclass(vllm.outputs.CompletionOutput)
+
+class _RequestOutput_tc(TensorClass["nocast"]):
+    request_id: str
+    prompt: str
+    prompt_token_ids: str
+    prompt_logprobs: str
+    outputs: str
+    finished: str
+    metrics: str
+    lora_request: str
+    encoder_prompt: str
+    encoder_prompt_token_ids: str
+    num_cached_tokens: str
+
+    def __post_init__(self):
+
+        def postproc(output):
```

Comment:
I've been in general pretty unopinionated about coding style, but here you are touching a pet peeve of mine so it's my turn to chime in :P I absolutely hate nested functions in Python, with a burning passion. I think that they make the code far less readable, and I go through literally anything and everything to avoid them. My usual go-to is to simply make a utils file somewhere else and put these functions there, but I'm happy to go with what the majority thinks here and with what's standard in Tune.

That said, where I put the flag is that at no point is double-nesting acceptable. Period, lol. This is unholy, come on! :D

Reply:
@vmoens I think this is you lol

Reply:
Typically, if we move these out (which we should!) i'd do it once the free functions are factored out in a separate file